signature-v4-parser_test.go

```diff
@@ -0,0 +1,427 @@
+		if testCase.expectedErrCode != actualErrCode {
+			t.Errorf("Test %d: Expected the APIErrCode to be %d, got %d", i+1, testCase.expectedErrCode, actualErrCode)
+		}
+
+	}
+}
+
+// TestParseSignature - validates the logic for extracting the signature string.
+func TestParseSignature(t *testing.T) {
+	testCases := []struct {
+		inputSignElement string
+		expectedSignStr  string
+		expectedErrCode  APIErrorCode
+	}{
+		// Test case - 1.
+		// SignElemenet doesn't have 2 parts on an attempt to split at '='.
+		// ErrMissingFields expected.
+		{
+			inputSignElement: "Signature",
+			expectedSignStr:  "",
+			expectedErrCode:  ErrMissingFields,
+		},
+		// Test case - 2.
+		// SignElemenet with missing "SignatureTag",ErrMissingSignTag expected.
+		{
+			inputSignElement: "Sign=",
+			expectedSignStr:  "",
+			expectedErrCode:  ErrMissingSignTag,
+		},
+		// Test case - 3.
+		// Test case with valid inputs.
+		{
+			inputSignElement: "Signature=abcd",
+			expectedSignStr:  "abcd",
+			expectedErrCode:  ErrNone,
+		},
+	}
+	for i, testCase := range testCases {
+		actualSignStr, actualErrCode := parseSignature(testCase.inputSignElement)
+		if testCase.expectedSignStr != actualSignStr {
+			t.Errorf("Test %d: Expected the result to be \"%s\", but got \"%s\". ", i+1, testCase.expectedSignStr, actualSignStr)
+
+		}
+		if testCase.expectedErrCode != actualErrCode {
+			t.Errorf("Test %d: Expected the APIErrCode to be %d, got %d", i+1, testCase.expectedErrCode, actualErrCode)
+		}
+
+	}
+}
+
+// TestParseSignedHeaders - validates the logic for extracting the signature string.
+func TestParseSignedHeaders(t *testing.T) {
+	testCases := []struct {
+		inputSignElement      string
+		expectedSignedHeaders []string
+		expectedErrCode       APIErrorCode
+	}{
+		// Test case - 1.
+		// SignElemenet doesn't have 2 parts on an attempt to split at '='.
+		// ErrMissingFields expected.
+		{
+			inputSignElement:      "SignedHeaders",
+			expectedSignedHeaders: nil,
+			expectedErrCode:       ErrMissingFields,
+		},
+		// Test case - 2.
+		// SignElemenet with missing "SigHeaderTag",ErrMissingSignHeadersTag expected.
+		{
+			inputSignElement:      "Sign=",
+			expectedSignedHeaders: nil,
+			expectedErrCode:       ErrMissingSignHeadersTag,
+		},
+		// Test case - 3.
+		// Test case with valid inputs.
+		{
+			inputSignElement:      "SignedHeaders=host;x-amz-content-sha256;x-amz-date",
+			expectedSignedHeaders: []string{"host", "x-amz-content-sha256", "x-amz-date"},
+			expectedErrCode:       ErrNone,
+		},
+	}
+
+	for i, testCase := range testCases {
+		actualSignedHeaders, actualErrCode := parseSignedHeaders(testCase.inputSignElement)
+		if strings.Join(testCase.expectedSignedHeaders, ",") != strings.Join(actualSignedHeaders, ",") {
+			t.Errorf("Test %d: Expected the result to be \"%v\", but got \"%v\". ", i+1, testCase.expectedSignedHeaders, actualSignedHeaders)
+
+		}
+		if testCase.expectedErrCode != actualErrCode {
```

Comment:
shouldn't this be validated first? becasue 'actualSignedHeaders' would be 'nil' for actualErrCode is not ErrNone.